### PR TITLE
UI: Add version release notes

### PR DESCRIPTION
**Pull request by kylesuss**
# This depends on https://github.com/storybookjs/frontpage/pull/107

Issue: #11232 

## What I did

This PR sets up the release notes logic in the build process & also in the ui.

## How to test

- **Is this testable with Jest or Chromatic screenshots?**

Yeah. I added some stories for the release notes page states, though I left the iframe out as that will likely change quite often.

- **Does this need a new example in the kitchen sink apps?**

Not sure. I added a new CLI flag for disabling the auto-messaging around release notes. Does that warrant a new example?

- **Does this need an update to the documentation?**

There is a new CLI flag, so that should be documented. Does that happen automatically though? I couldn't quite tell. I added it to `lib/core/src/server/cli/dev.js` with a note, which seemed to be documentation, but not sure where that shows up.

### CODE REVIEW

File: lib/api/src/modules/settings.ts
Comment (tmeasday) [inline lib/api/src/modules/settings.ts:19]: Does this have any effect? I didn't think that actually led to the first story being selected any more..

File: lib/api/src/modules/stories.ts
Comment (tmeasday) [inline lib/api/src/modules/stories.ts:280]: Should this be more general? I can't imagine it is a good idea to close any settings page randomly when the SB loads.

File: lib/api/src/modules/url.ts
Comment (tmeasday) [inline lib/api/src/modules/url.ts:169]: What's the idea behind moving this? I can't say I fully understand why this is where it is (or what's up with the `setTimeout(..., 1)`s, but can you explain why it's moving?

File: lib/api/src/modules/settings.ts
Comment (tmeasday) [inline lib/api/src/modules/settings.ts:32]: I guess this means we decided we'd use URLs for settings after all?

File: lib/api/src/modules/settings.ts
Comment (tmeasday) [reply lib/api/src/modules/settings.ts:19]: I'm not sure what the best way to do it otherwise is 🤔

File: lib/api/src/modules/release-notes.ts
Comment (tmeasday) [inline lib/api/src/modules/release-notes.ts:67]: I thought we were going to keep things simple and not track any persistent state in the browser. Did that change?

Comment (kylesuss) [reply lib/api/src/modules/release-notes.ts:67]: I wanted to, but when it came down to implementation, I wasn't sure how to do that & prevent the release notes from re-opening every time you re-loaded your Storybook (a browser refresh). It seemed like I needed to track that on the client to prevent them from getting "redirected" to the release notes every time they refreshed the page. Do you have any ideas on how to prevent doing this?

File: lib/api/src/modules/settings.ts
Comment (kylesuss) [reply lib/api/src/modules/settings.ts:32]: Yeah. I mean -- I'm quite neutral about it. I figured if the experience would work & be linkable as well, that it wouldn't hurt to keep the URLs around. Any thoughts on that?

File: lib/api/src/modules/url.ts
Comment (kylesuss) [reply lib/api/src/modules/url.ts:169]: The whole `setTimeout` thing feels like a hack. TBH I am not sure why it is being done that way. I suspect it is some workaround for a timing issue related to postMessage events between the manager & preview. That said, I moved this stuff around because the `initialUrlSupport` function was triggering redirects _before_ all of the "modules" have been setup and therefore, I didn't have access to the state of the `release_notes` module to know if that required a redirect. By moving it, I could `await` the `release_notes` module & know whether or not we should redirect there.

File: lib/api/src/modules/stories.ts
Comment (kylesuss) [reply lib/api/src/modules/stories.ts:280]: Yeah, it would be nice to generalize. I think the hard part about that is that this callback handles all events for `CURRENT_STORY_WAS_SET` and therefore I can't just say "if on the settings page, don't redirect" because you might be intentionally navigating away from the settings page. Thoughts?

File: lib/api/src/modules/settings.ts
Comment (kylesuss) [reply lib/api/src/modules/settings.ts:19]: It has the effect of closing the settings page and showing this screen:

Screen Shot 2020-06-30 at 11 26 49 AM

I would have liked to navigate to the first story & spent some time trying to figure out how to do that but it seemed way more complicated than this. But, if anyone knows how to do that, I can change it up.

File: lib/api/src/modules/url.ts
Comment (kylesuss) [reply lib/api/src/modules/url.ts:169]: That said, I am slightly confused as there are a bunch of redirects happening & given what I had to do here:

https://github.com/storybookjs/storybook/pull/11360/files#r447366304

There doesn't seem to be a dedicated place to isolate this logic.

File: lib/api/src/modules/settings.ts
Comment (tmeasday) [reply lib/api/src/modules/settings.ts:19]: We could add a `selectFirstStory` method to the stories module. 

How do you get into this state?

File: lib/api/src/modules/stories.ts
Comment (tmeasday) [reply lib/api/src/modules/stories.ts:280]: Should we make it the direct action that closes the settings, rather than the side effect? 

E.g close the settings when you click the close button or a story in the sidebar, instead of when the story change happens in the sidebar. 

I actually think this’ll happen pretty naturally if you block the routing here when settings is open. Here’s my thinking. There are two ways this event can fire:

1. Preview boots (at the very start OR on hot module reload)

2. In response to the manager selecting a new story

This code is here only for case 1, as:

If we break down what happens in 2:

A. Something triggers `SELECT_STORY`
B. That makes the manager change route (this would close settings)
C. Manager emits `SET_CURRENT_STORY` in response to route change
D. Preview changes story, emits `CURRENT_STORY_WAS_SET` which hits this code path
E. We try and route again to where we already are, which is a no-op

It’s a bit confusing for sure but hopefully that helps?

File: lib/api/src/modules/settings.ts
Comment (tmeasday) [reply lib/api/src/modules/settings.ts:32]: Sure, I’m not wedded to either approach.

File: lib/api/src/modules/release-notes.ts
Comment (tmeasday) [reply lib/api/src/modules/release-notes.ts:67]: Ahh, got it. Yeah that makes sense. Thanks for the clarification

File: lib/api/src/modules/settings.ts
Comment (kylesuss) [reply lib/api/src/modules/settings.ts:19]: You get into this state by triggering the release notes to open immediately when you start Storybook. When you do that, you don't have any story history & therefore there is no story to navigate "back" to.

Comment (kylesuss) [reply lib/api/src/modules/settings.ts:19]: `selectFirstStory` seems reasonable -- the thing I was struggling with was getting the state of the stories in order. There was a `storiesHash` but it wasn't sorted. I'm looking now at `jumpToStory` and think I can copy that logic to sort & find the first story.

File: lib/core/src/server/build-dev.js
Comment (kylesuss) [inline lib/core/src/server/build-dev.js:147]: Of note here is that I decided to pass the cache in so it would be easier to mock when testing. I can never figure out how to mock a node module w/ jest & change the return value in each test.

File: lib/api/src/modules/settings.ts
Comment (kylesuss) [reply lib/api/src/modules/settings.ts]: Updated & got `selectFirstStory` to work 👌

File: lib/api/src/modules/url.ts
Comment (kylesuss) [reply lib/api/src/modules/url.ts]: @tmeasday I moved these redirects back for now and commented them out. I was planning to take another look at how it affects things once you've had the chance to look into removing them.

File: lib/api/src/modules/stories.ts
Comment (kylesuss) [reply lib/api/src/modules/stories.ts]: Updated w/ a reference to the new settings module to determine if it is open. Seems to work well (though I have all of those `setTimeout` navigate calls commented out) -- I just need to circle back once those are removed.

File: lib/core/src/server/build-dev.js
Comment (lifeiscontent) [reply lib/core/src/server/build-dev.js:147]: @kylesuss have you tried https://jestjs.io/docs/en/manual-mocks ?

File: lib/api/src/modules/url.ts
Comment (kylesuss) [inline lib/api/src/modules/url.ts:94]: Some tests are failing because of this, which I expect.

File: lib/core/src/server/build-dev.js
Comment (kylesuss) [reply lib/core/src/server/build-dev.js:147]: I didn't. Mostly a lot of combinations of `jest.mock`, `jest.doMock`, `jest.mockReturnValue`, etc.

> If the module you are mocking is a Node module (e.g.: lodash), the mock should be placed in the __mocks__ directory adjacent to node_modules (unless you configured roots to point to a folder other than the project root) and will be automatically mocked. There's no need to explicitly call jest.mock('module_name').

Do we want to mock this module all the time? That seems a bit heavy handed.

Comment (tmeasday) [reply lib/core/src/server/build-dev.js:147]: @kylesuss something like

```js
import foo from 'bar';

jest.mock('bar');

it('...', () => {
  foo.mockReturnValue('...');
});
```

Is that what you mean?

File: lib/api/src/modules/release-notes.ts
Comment (tmeasday) [inline lib/api/src/modules/release-notes.ts:51]: Is there a reason this logic needs to be here and not inside `showReleaseNotesOnLaunch`?

File: lib/api/src/modules/release-notes.ts
Comment (tmeasday) [inline lib/api/src/modules/release-notes.ts:55]: Maybe we should just write this when the release notes page renders?

I think that would avoid the issue you are running into here with the `async`-ness

File: lib/api/src/modules/release-notes.ts
Comment (tmeasday) [inline lib/api/src/modules/release-notes.ts:33]: This is sort of an unusual pattern. I feel like it's pretty confusing to have a `async` function called from a module's `initModule` given that we don't actually `await` initModule...

File: lib/api/src/modules/stories.ts
Comment (tmeasday) [inline lib/api/src/modules/stories.ts:205]: Can we just use a `find` here rather than filtering the whole list?

File: lib/api/src/modules/stories.ts
Comment (tmeasday) [inline lib/api/src/modules/stories.ts:213]: This doesn't do anything, I don't think. If there's no stories at all I think should just set the URL to `/`.

**Review by tmeasday — CHANGES_REQUESTED** [review]
Some comments about modules

File: lib/core/src/server/build-dev.js
Comment (kylesuss) [reply lib/core/src/server/build-dev.js:147]: Yeah I tried that.

```js
import Cache from 'file-system-cache';
import { getReleaseNotesData, RELEASE_NOTES_CACHE_KEY } from './build-dev';

jest.mock('file-system-cache');

it('handles errors gracefully', async () => {
  Cache.mockReturnValue(() => ({}));
  // The rest of the test
});
```

The `cache` in the module is `undefined`.

This works:

```js
jest.mock('file-system-cache', () => ({
  __esModule: true,
  default: () => ({}),
}));
```

But it means every test will have the same return value. I can't seem to override it with `mockReturnValue` inside the test.

Jest's docs [suggest](https://jestjs.io/docs/en/jest-object#jestdomockmodulename-factory-options) using `doMock` in these cases for ES6 imports so I tried this:

```js
it('handles errors gracefully', async () => {
    const version = '4.0.0';
    jest.doMock('file-system-cache', () => ({
      __esModule: true,
      default: jest.fn(() => ({})),
    }));

    return import('file-system-cache').then(async () => {
      expect(await getReleaseNotesData(version)).toEqual({
        currentVersion: version,
        showOnFirstLaunch: false,
        success: false,
      });
    });
  });
```

But `cache` doesn't end up having the mock value -- it is just the module's normal FileSystemCache object.

So I just tried manual mocks as suggested by @lifeiscontent:

```js
// __mocks__/file-system-cache.js
module.exports = jest.fn(() => ({ a: 'b' }));

// build-dev.test.js
it('handles errors gracefully', async () => {
  Cache.mockReturnValue(() => ({}));
  // The rest of the test
});
```

And the value is `({ a: 'b' })`.

Haha. So I donno -- maybe I am missing something but I have tried quite a lot.

File: lib/api/src/modules/stories.ts
Comment (kylesuss) [reply lib/api/src/modules/stories.ts:205]: Yeah that makes sense 👍  I just blindly copied the logic from the other function.

File: lib/api/src/modules/stories.ts
Comment (kylesuss) [reply lib/api/src/modules/stories.ts]: https://github.com/storybookjs/storybook/pull/11360/commits/5e1e0908668e0f4d50ecfe11afc7e29e1b7d86d3

File: lib/api/src/modules/stories.ts
Comment (kylesuss) [reply lib/api/src/modules/stories.ts]: https://github.com/storybookjs/storybook/pull/11360/commits/5e1e0908668e0f4d50ecfe11afc7e29e1b7d86d3

File: lib/api/src/modules/release-notes.ts
Comment (kylesuss) [reply lib/api/src/modules/release-notes.ts]: Hmm. Good point. I guess not. Moving it there means I have no logic inside of `initModule`, which feels kind of weird but maybe that's fine. I was trying to follow a pattern of defining the state inside of `initModule` but only because I thought that was how the other modules were doing it. I went ahead and moved it to `showReleaseNotesOnLaunch` 👌

File: lib/api/src/modules/stories.ts
Comment (tmeasday) [reply lib/api/src/modules/stories.ts]: Heh yeah I figured :)

**Review by tmeasday — APPROVED** [review]
LGTM once the timeouts pr is merged!

File: lib/ui/src/settings/about_page.tsx
Comment (shilman) [inline lib/ui/src/settings/about_page.tsx:36]: since this was written we've switched to named exports (as opposed to default exports), so please export it as AboutPage and update the reference.

File: lib/ui/src/settings/shortcuts_page.tsx
Comment (shilman) [inline lib/ui/src/settings/shortcuts_page.tsx:22]: use named exports

File: lib/api/src/modules/url.ts
Comment (tmeasday) [reply lib/api/src/modules/url.ts]: My PR is merged now!

File: lib/api/src/modules/url.ts
Comment (kylesuss) [reply lib/api/src/modules/url.ts]: No longer relevant!

File: lib/api/src/modules/url.ts
Comment (kylesuss) [reply lib/api/src/modules/url.ts]: Nice. I merged that in 👌

**Comment by kylesuss** [comment]
@tmeasday @shilman ok this is up to date w/ @tmeasday 's changes for the URL module & the named export stuff has been addressed. Let me know what else needs addressing & I can handle tomorrow.

**Review by tmeasday — APPROVED** [review]
LGTM @kylesuss 

I guess you should approve the Chromatic changes.

Also I am not sure what's up with CI, you might want to try merging latest `next`.

**Comment by kylesuss** [comment]
@tmeasday @shilman this seems good to go, pending UI Review. Feel free to merge once https://github.com/storybookjs/frontpage/pull/107 is merged.

**Pull request merged**
